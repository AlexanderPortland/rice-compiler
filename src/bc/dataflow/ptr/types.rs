use std::{collections::HashMap, fmt, sync::Arc};

use indexical::{ArcIndexSet, ArcIndexVec, IndexedDomain, set::IndexSet};
use internment::Intern;
use serde::Serialize;

use crate::{
    bc::{
        types::{
            AllocArgs, AllocKind, Function, Local, LocalIdx, Location, Operand, Place,
            ProjectionElem, Rvalue, Type,
        },
        visit::Visit,
    },
    interned,
};

/// A location in memory.
#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
pub enum MemLoc {
    /// A normal value stored in a [Local] (on the stack).
    Local(LocalIdx),
    /// A tuple value within an [Allocation]. It's conceptually on the heap despite that
    /// optimizations might decide to actually allocate it on the stack instead.
    Allocated(Allocation, AllocProj),
}

indexical::define_index_type! {
    pub struct MemLocIdx for MemLoc = u32;
}

impl std::fmt::Display for MemLoc {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MemLoc::Local(l) => <LocalIdx as std::fmt::Display>::fmt(l, f),
            MemLoc::Allocated(..) => <Self as std::fmt::Debug>::fmt(self, f),
        }
    }
}

#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy, Serialize, PartialOrd, Ord)]
pub enum AllocProj {
    Index,
    Field(usize),
}

/// An allocation, identified by the location at which it was allocated.
#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
pub enum Allocation {
    /// An allocation passed into the function as an argument.
    FromArg(PtrPlace),
    /// A 'real' allocation made within the current function at a given location.
    Real(Location),
}

indexical::define_index_type! {
    pub struct AllocationIdx for Allocation = u32;
}

impl Allocation {
    pub fn from_arg(arg_p: impl Into<PtrPlace>) -> Self {
        Self::FromArg(arg_p.into())
    }

    pub fn from_loc(loc: Location) -> Self {
        Self::Real(loc)
    }

    pub fn with_index_proj(&self, proj: &AllocProj) -> MemLoc {
        MemLoc::Allocated(*self, *proj)
    }
}

/// A simplified version of places that ensure all indexes are identical and
/// projection types don't matter
#[derive(Debug, Hash, PartialEq, Eq, Clone, Serialize, PartialOrd, Ord)]
pub struct PtrPlaceData {
    pub local: LocalIdx,
    pub projection: Vec<AllocProj>,
}

interned!(PtrPlace, PtrPlaceData);

impl PtrPlace {
    pub fn extend_projection(self, proj: AllocProj) -> Self {
        let mut d = (*self.0).clone();
        d.projection.push(proj);
        PtrPlace(Intern::new(d))
    }
}

impl From<Place> for PtrPlace {
    fn from(value: Place) -> Self {
        PtrPlace(Intern::new(PtrPlaceData {
            local: value.local,
            projection: value
                .projection
                .iter()
                .map(|proj_el| match proj_el {
                    ProjectionElem::Field { index, ty } => AllocProj::Field(*index),
                    ProjectionElem::Index { index, ty } => AllocProj::Index,
                })
                .collect(),
        }))
    }
}

impl fmt::Display for PtrPlace {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.local)?;
        for proj in &self.projection {
            match proj {
                AllocProj::Field(index) => write!(f, ".{index}")?,
                AllocProj::Index => write!(f, "[?]")?,
            }
        }
        Ok(())
    }
}
