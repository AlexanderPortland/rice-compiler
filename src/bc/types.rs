//! Type definitions for the bytecode.

use std::{borrow::Borrow, collections::HashMap, sync::Arc};

use either::Either;
use indexical::IndexedDomain;
use internment::Intern;
use itertools::Itertools;
use petgraph::{
    Direction,
    graph::{DiGraph, NodeIndex},
};
use serde::Serialize;
use smallvec::{SmallVec, smallvec};
use strum::Display;

use crate::{
    ast::types::{Annotation, Span},
    interned,
    utils::{self, Symbol},
};

pub use crate::tir::types::{Binop, Const, Type, TypeKind};

#[derive(Debug, Clone, Serialize)]
pub struct Program(Vec<Function>);

impl Program {
    #[must_use]
    pub fn new(funcs: Vec<Function>) -> Self {
        Program(funcs)
    }

    #[must_use]
    pub fn functions(&self) -> &[Function] {
        &self.0
    }

    /// NOTE: returns none if the function's defined within the standard library.
    pub fn find_function(&self, sym: impl Borrow<Symbol>) -> Option<&Function> {
        let sym = sym.borrow();
        let matching = self
            .functions()
            .into_iter()
            .filter(|func| func.name == *sym)
            .collect::<Vec<_>>();

        if matching.is_empty() {
            return None;
        }

        assert!(
            matching.len() == 1,
            "we should only have a single function named, {sym}"
        );
        Some(matching[0])
    }

    pub fn main_function(&self) -> &Function {
        self.find_function(Symbol::main())
            .expect("we should have a main function")
    }

    pub fn functions_mut(&mut self) -> &mut [Function] {
        &mut self.0
    }

    #[must_use]
    pub fn into_functions(self) -> Vec<Function> {
        self.0
    }
}

#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug, Serialize)]
pub struct Local {
    /// The type of the local variable.
    pub ty: Type,

    /// A source-level name for the local variable, if it exists.
    pub name: Option<Symbol>,
}

indexical::define_index_type! {
    /// An index for locals, for use in indexical data structures.
    pub struct LocalIdx for Local = u32;
}

#[derive(Debug, Clone, Serialize)]
pub struct Function {
    /// The global name of the function. Must be unique.
    pub name: Symbol,

    /// The first `num_param` locals are parameters.
    pub num_params: usize,

    /// All the local variables in the function.
    pub locals: Arc<IndexedDomain<Local>>,

    /// The return type.
    pub ret_ty: Type,

    /// The control-flow graph of instructions.
    pub body: Body,

    /// Source-level annotations attached to this function.
    pub annots: Vec<Annotation>,
}

impl Function {
    /// Returns an iterator over the local index and type of each parameter.
    pub fn params(&self) -> impl Iterator<Item = (LocalIdx, Type)> {
        (0..self.num_params).map(|i| {
            let local = LocalIdx::from_usize(i);
            (local, self.locals.value(local).ty)
        })
    }

    /// Returns true if the function is annotated with `#[jit]`.
    #[must_use]
    pub fn jit(&self) -> bool {
        self.annots.iter().any(|annot| annot.name == "jit")
    }

    /// Returns true if the function is annotated with `#[secure]`.
    #[must_use]
    pub fn secure(&self) -> bool {
        self.annots.iter().any(|annot| annot.name == "secure")
    }
}

pub type Cfg = DiGraph<BasicBlock, ()>;

#[derive(Debug, Clone, Default, Serialize)]
pub struct Body {
    /// The instructions in the body.    
    cfg: Cfg,

    /// A set of all locations in the body, useful for maps indexed by location.
    locations: Arc<IndexedDomain<Location>>,

    /// Reverse-postorder traversal of basic blocks in the CFG.    
    #[serde(skip)]
    rpo: Vec<BasicBlockIdx>,
}

impl Body {
    pub fn new(cfg: Cfg) -> Self {
        let rpo = utils::reverse_post_order(&cfg, Location::START.block.into())
            .into_iter()
            .map(BasicBlockIdx::from)
            .collect_vec();

        let locations = Self::build_domain(&cfg, &rpo);

        Body {
            cfg,
            locations,
            rpo,
        }
    }

    fn build_domain(cfg: &Cfg, rpo: &[BasicBlockIdx]) -> Arc<IndexedDomain<Location>> {
        Arc::new(
            rpo.iter()
                .copied()
                .flat_map(|block| {
                    // println!("building domain for block {:?}", block);
                    let num_instrs = cfg.node_weight(block.into()).unwrap().statements.len() + 1;
                    (0..num_instrs).map(move |instr| Location { block, instr })
                })
                .collect(),
        )
    }

    fn block_is_redundant(&self, block_no: BasicBlockIdx, min_connections: usize) -> bool {
        self.cfg()
            .neighbors_directed(block_no.into(), petgraph::Direction::Incoming)
            .count()
            <= min_connections
            && block_no != 0
    }

    /// Shadow remove a basic block from this body's structure. Removes all nodes that it points to
    /// but doesn't actually remove the block itself to preserve indexing. But nothing will point to it so doesn't really matter.
    pub fn remove_block_cfg_edges(&mut self, block_no: BasicBlockIdx) {
        // println!("remove {:?}", block_no);
        assert_eq!(
            self.cfg()
                .neighbors_directed(block_no.into(), petgraph::Direction::Incoming)
                .count(),
            0
        );
        assert_ne!(block_no, 0, "dont try to remove the first block you dofus");

        // clear all the statments, as we'll remove this block
        self.data_mut(block_no).statements.clear();

        // remove all neighbors in the cfg
        for out_neighbor in self
            .cfg()
            .neighbors_directed(block_no.into(), petgraph::Direction::Outgoing)
            .collect::<Vec<_>>()
        {
            // if this neighbor only has one incoming connection (us), we want to remove them too...

            let edge = self
                .cfg()
                .find_edge(block_no.into(), out_neighbor)
                .expect("should have an edge to all neighbors");
            self.cfg_mut()
                .remove_edge(edge)
                .expect("should have an edge here..");

            if self.block_is_redundant(out_neighbor.into(), 0) {
                self.remove_block_cfg_edges(out_neighbor.into());
            }
        }
    }

    /// Regenerates the rpo, domain and locations. For use when adding or remove basic blocks.
    pub fn regenerate_everything(&mut self) {
        // I think just reconstructing it should work...
        *self = Self::new(std::mem::take(&mut self.cfg));
    }

    /// Regenerates the location domain, to be used after adding or removing instructions.
    pub fn regenerate_domain(&mut self) {
        self.locations = Self::build_domain(&self.cfg, &self.rpo);
    }

    /// Returns the data corresponding to the given basic block.
    #[must_use]
    pub fn data(&self, node: BasicBlockIdx) -> &BasicBlock {
        self.cfg.node_weight(node.into()).unwrap()
    }

    /// Returns a mutable handle to the data in a given basic block.
    pub fn data_mut(&mut self, node: BasicBlockIdx) -> &mut BasicBlock {
        self.cfg.node_weight_mut(node.into()).unwrap()
    }

    /// Returns an iterator over basic block indices.
    ///
    /// Guaranteed to be a reverse-postorder.
    #[must_use]
    pub fn blocks(&self) -> impl DoubleEndedIterator<Item = BasicBlockIdx> {
        self.rpo.iter().copied()
    }

    /// Returns the underlying CFG.
    #[must_use]
    pub fn cfg(&self) -> &Cfg {
        &self.cfg
    }

    /// Returns the underlying CFG.
    pub fn cfg_mut(&mut self) -> &mut Cfg {
        &mut self.cfg
    }

    /// Returns the location domain.
    #[must_use]
    pub fn locations(&self) -> &Arc<IndexedDomain<Location>> {
        &self.locations
    }

    /// Returns the instruction at a given location.
    #[must_use]
    pub fn instr(&self, loc: Location) -> Either<&Statement, &Terminator> {
        self.data(loc.block).get(loc.instr)
    }

    /// Returns a mutable handle to the instruction at a given locatin.
    pub fn instr_mut(&mut self, loc: Location) -> Either<&mut Statement, &mut Terminator> {
        self.data_mut(loc.block).get_mut(loc.instr)
    }

    /// Returns all locations which can possibly come after the given location.
    #[must_use]
    pub fn successors(&self, loc: Location) -> SmallVec<[Location; 2]> {
        if loc.instr == self.data(loc.block).terminator_index() {
            self.cfg
                .neighbors_directed(loc.block.into(), Direction::Outgoing)
                .map(|block| Location {
                    block: BasicBlockIdx::from(block),
                    instr: 0,
                })
                .collect()
        } else {
            smallvec![Location {
                block: loc.block,
                instr: loc.instr + 1,
            }]
        }
    }

    /// Returns all locations which can possibly come before the given location.
    #[must_use]
    pub fn predecessors(&self, loc: Location) -> SmallVec<[Location; 2]> {
        if loc.instr == 0 {
            self.cfg
                .neighbors_directed(loc.block.into(), Direction::Incoming)
                .map(|block| {
                    let block = BasicBlockIdx::from(block);
                    let instr = self.data(block).terminator_index();
                    Location { block, instr }
                })
                .collect()
        } else {
            smallvec![Location {
                block: loc.block,
                instr: loc.instr - 1
            }]
        }
    }
}

index_vec::define_index_type! {
    pub struct BasicBlockIdx = u32;
}

impl BasicBlockIdx {
    /// Returns the location of the first instruction in the basic block.
    #[must_use]
    pub fn entry(self) -> Location {
        Location {
            block: self,
            instr: 0,
        }
    }
}

impl From<NodeIndex> for BasicBlockIdx {
    fn from(value: NodeIndex) -> Self {
        BasicBlockIdx::new(value.index())
    }
}

impl From<BasicBlockIdx> for NodeIndex {
    fn from(value: BasicBlockIdx) -> Self {
        NodeIndex::new(value.index())
    }
}

/// A coordinate for a particular instruction in a CFG.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize)]
pub struct Location {
    /// The basic block containing the instruction.
    pub block: BasicBlockIdx,

    /// The index of the instruction within the basic block.
    pub instr: usize,
}

indexical::define_index_type! {
    pub struct LocationIdx for Location = u32;
}

impl Location {
    /// The location of the starting instruction in any CFG.
    pub const START: Location = Location {
        block: BasicBlockIdx { _raw: 0 },
        instr: 0,
    };
}

/// A basic block in the control flow graph, containing a sequence of statements followed by a terminator.
#[derive(Debug, Clone, Serialize)]
pub struct BasicBlock {
    /// Statements executed in sequence.
    pub statements: Vec<Statement>,

    /// The final instruction, which can branch to other basic blocks.
    pub terminator: Terminator,
}

impl BasicBlock {
    /// Get the ith instruction in the basic block.
    #[must_use]
    pub fn get(&self, i: usize) -> Either<&Statement, &Terminator> {
        assert!(i <= self.statements.len());
        if i == self.statements.len() {
            Either::Right(&self.terminator)
        } else {
            Either::Left(&self.statements[i])
        }
    }

    // Get a mutable handle to the ith instruction in the basic block.
    pub fn get_mut(&mut self, i: usize) -> Either<&mut Statement, &mut Terminator> {
        assert!(i <= self.statements.len());
        if i == self.statements.len() {
            Either::Right(&mut self.terminator)
        } else {
            Either::Left(&mut self.statements[i])
        }
    }

    #[must_use]
    pub fn terminator_index(&self) -> usize {
        self.statements.len()
    }
}

/// An instruction of the form `place = rvalue`.
#[derive(Debug, Clone, Serialize)]
pub struct Statement {
    /// The place in memory being assigned to.
    pub place: Place,

    /// An expression that generates a value to set to the place.
    pub rvalue: Rvalue,

    /// A best-effort source-level span for the statement.
    pub span: Span,
}

/// An argument to an [`Rvalue`].
#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize)]
pub enum Operand {
    Const(Const),
    Place(Place),
    Func { f: Symbol, ty: Type },
}

impl Operand {
    #[must_use]
    pub fn ty(&self) -> Type {
        match self {
            Operand::Place(p) => p.ty,
            Operand::Const(c) => c.ty(),
            Operand::Func { ty, .. } => *ty,
        }
    }

    #[must_use]
    pub fn as_place(&self) -> Option<Place> {
        match self {
            Operand::Place(p) => Some(*p),
            _ => None,
        }
    }
}

/// The kind of object allocated by an [`Rvalue::Alloc`] instruction.
#[derive(Debug, Clone, Copy, Display, Serialize)]
pub enum AllocKind {
    Struct,
    Tuple,
    Array,
}

/// The location in memory of an object allocated by an [`Rvalue::Alloc`] instruction.
#[derive(Debug, Clone, Copy, Display, Serialize, PartialEq, Eq, PartialOrd, Ord)]
// #[serde(rename_all = "lowercase")]
#[serde(into = "String")]
pub enum AllocLoc {
    Stack,
    Heap,
}

impl From<AllocLoc> for String {
    fn from(value: AllocLoc) -> Self {
        format!("{value}")
    }
}

/// The inputs to an allocation by an [`Rvalue::Alloc`] instruction.
#[derive(Debug, Clone, Display, Serialize)]

pub enum AllocArgs {
    /// A fixed-size literal, e.g. `(x, y)` or `[x, y]`.
    Lit(Vec<Operand>),
    Repeated {
        op: Operand,
        count: Operand,
    },
}

/// An expression that is assigned to a [`Place`] in a [`Statement`].
#[derive(Debug, Clone, Serialize)]
pub enum Rvalue {
    /// The direct value of an operand, e.g. `1` or `x`.
    Operand(Operand),

    /// Cast another operand to a new type, e.g. `x as float`.
    Cast { op: Operand, ty: Type },

    /// A first-class function with a (potentially-empty) environment, e.g., `closure(lambda, [x, y])`.
    Closure { f: Symbol, env: Vec<Operand> },

    /// An allocation of a data structure larger than a register.
    Alloc {
        kind: AllocKind,
        loc: AllocLoc,
        args: AllocArgs,
    },

    /// A binary operator.
    Binop {
        op: Binop,
        left: Operand,
        right: Operand,
    },

    /// A call to a non-method function.
    Call { f: Operand, args: Vec<Operand> },

    /// A call to a method function w/ dynamic dispatch.
    MethodCall {
        /// The object containing a vtable.
        receiver: Operand,

        /// The index of the vtable to lookup.
        method: usize,

        /// The arguments to the method.
        args: Vec<Operand>,
    },
}

impl Rvalue {
    /// Returns a vector of all the places used in the rvalue.
    #[must_use]
    pub fn places(&self) -> SmallVec<[Place; 2]> {
        match self {
            Rvalue::Operand(op) | Rvalue::Cast { op, .. } => op.as_place().into_iter().collect(),
            Rvalue::Alloc { args, .. } => match args {
                AllocArgs::Lit(ops) => ops.iter().filter_map(Operand::as_place).collect(),
                AllocArgs::Repeated { op, count } => vec![op.as_place(), count.as_place()]
                    .into_iter()
                    .flatten()
                    .collect(),
            },
            Rvalue::Call { args: ops, .. } | Rvalue::Closure { env: ops, .. } => {
                ops.iter().filter_map(Operand::as_place).collect()
            }
            Rvalue::MethodCall { receiver, args, .. } => args
                .iter()
                .chain([receiver])
                .filter_map(Operand::as_place)
                .collect(),
            Rvalue::Binop { left, right, .. } => left
                .as_place()
                .into_iter()
                .chain(right.as_place())
                .collect(),
        }
    }
}

/// The final instruction in a basic block.
#[derive(Debug, Clone, Serialize)]
pub struct Terminator {
    pub kind: TerminatorKind,
    pub span: Span,
}

impl Terminator {
    #[must_use]
    pub fn kind(&self) -> &TerminatorKind {
        &self.kind
    }

    pub fn kind_mut(&mut self) -> &mut TerminatorKind {
        &mut self.kind
    }

    pub fn places(&self) -> SmallVec<[Place; 2]> {
        match &self.kind {
            TerminatorKind::Return(op) | TerminatorKind::CondJump { cond: op, .. } => {
                op.as_place().into_iter().collect()
            }
            TerminatorKind::Jump(_) => SmallVec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub enum TerminatorKind {
    /// Return the value of the operand from the function.
    Return(Operand),

    /// Unconditionally jump to the given basic block.
    Jump(BasicBlockIdx),

    /// Conditionally jump to `true_` if `cond` is true, and jump to `false_` otherwise.
    CondJump {
        cond: Operand,
        true_: BasicBlockIdx,
        false_: BasicBlockIdx,
    },
}

impl Terminator {
    /// Remap the basic blocks inside the terminator, used during CFG construction.
    pub fn remap(&mut self, map: &HashMap<BasicBlockIdx, BasicBlockIdx>) {
        match &mut self.kind {
            TerminatorKind::Jump(block) => *block = map[block],
            TerminatorKind::CondJump { true_, false_, .. } => {
                *true_ = map[true_];
                *false_ = map[false_];
            }
            TerminatorKind::Return(..) => {}
        }
    }
}

/// A reference to a particular location in memory relative to a particular function.
#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize)]
pub struct PlaceData {
    /// The root of the reference, a local on a stack frame.
    pub local: LocalIdx,

    /// A sequence of projections off of the local.
    pub projection: Vec<ProjectionElem>,

    /// The type of the data at the location.
    pub ty: Type,
}

interned!(Place, PlaceData);

impl Place {
    #[must_use]
    pub fn new(local: LocalIdx, projection: Vec<ProjectionElem>, ty: Type) -> Self {
        Place(Intern::new(PlaceData {
            local,
            projection,
            ty,
        }))
    }

    #[must_use]
    pub fn extend_projection(
        self,
        elems: impl IntoIterator<Item = ProjectionElem>,
        ty: Type,
    ) -> Self {
        let mut projection = self.projection.clone();
        projection.extend(elems);
        Place::new(self.local, projection, ty)
    }
}

/// A lookup into a field of a data structure.
#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize)]
pub enum ProjectionElem {
    /// Get the field of a fixed-sized structure.
    Field { index: usize, ty: Type },

    /// Index into a fixed-size array.
    Index { index: Operand, ty: Type },
}
