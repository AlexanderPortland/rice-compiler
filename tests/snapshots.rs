use anyhow::Result;
use regex::Regex;
use rice::bc::types::AllocLoc;
use serde::Serialize;
use std::{collections::BTreeMap, fs, path::Path, process::Command};
use walkdir::WalkDir;

#[derive(Serialize)]
struct CompilerOutput {
    success: bool,
    stdout: String,
    stderr: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    stats: Option<CompilationStats>,
}

#[derive(Serialize)]
/// Statistics about our generated bytecode & its execution to prevent performance regressions.
struct CompilationStats {
    /// The total number of statements in our bytecode (including terminators)
    bc_stmts: usize,
    /// The number of locals in our bytecode
    bc_locals: usize,
    /// The number of basic blocks in our bytecode
    bc_blocks: usize,
    /// The total number of function calls when executing the program.
    exec_calls: usize,
    /// The total number of steps taken to execute the program
    exec_steps: usize,
    #[serde(skip_serializing_if = "BTreeMap::is_empty")]
    allocs: BTreeMap<AllocLoc, usize>,
}

fn compile(path: &Path, args: &str) -> Result<CompilerOutput> {
    let mut cmd = Command::new("./target/debug/rice");
    cmd.arg(path);
    cmd.args(shlex::split(args).unwrap());
    println!("running command {:?}", cmd);
    let output = cmd.output()?;
    let success = output.status.success();

    let stats = success.then(|| get_bytecode_stats(path, args));

    println!("stats got");

    Ok(CompilerOutput {
        success: output.status.success(),
        stats,
        stdout: String::from_utf8(output.stdout)?,
        stderr: String::from_utf8(output.stderr)?,
    })
}

/// Recompile with debug to check for stack allocations & loc
fn get_bytecode_stats(path: &Path, args: &str) -> CompilationStats {
    let mut cmd = Command::new("./target/debug/rice");
    cmd.arg(path);
    cmd.env("RUST_LOG", "rice=debug");
    cmd.args(shlex::split(args).unwrap());
    println!("running command {:?}", cmd);
    let output = cmd.output().expect("should run");

    let full_output_str = String::from_utf8(output.stderr).expect("shd");

    let start = full_output_str
        .as_str()
        .match_indices("] Optimized BC:")
        .last()
        .unwrap_or_else(|| {
            full_output_str
                .as_str()
                .match_indices("] Initial BC:")
                .last()
                .expect("should have A bytecode output")
        });
    let start_i = start.0 + start.1.len();

    let output_str = &full_output_str[start_i..];
    let output_str = &output_str[..output_str.find("\n[").unwrap_or(output_str.len())];

    let alloc_locs = [(AllocLoc::Heap, "Heap"), (AllocLoc::Stack, "Stack")];

    let re = Regex::new(r"RT:\s*\n?\s*calls:\s*(\d+),\s*steps:\s*(\d+)").unwrap();
    let caps = re.captures(&full_output_str).expect("should match RT");

    let rt_calls: usize = caps[1].parse().unwrap();
    let rt_steps: usize = caps[2].parse().unwrap();

    let bc_blocks = Regex::new("\n\\s*bb\\d+:")
        .unwrap()
        .find_iter(output_str)
        .count();

    CompilationStats {
        allocs: alloc_locs
            .iter()
            .filter_map(|(loc, name)| {
                let num = output_str.matches(&format!("@ {name}")).count();
                if num > 0 { Some((*loc, num)) } else { None }
            })
            .collect(),
        bc_locals: Regex::new(r"\n\s*local\s+x?\d+:")
            .unwrap()
            .find_iter(output_str)
            .count(),
        bc_blocks,
        // All statements have an equals sign, plus a terminator for each basic block.
        bc_stmts: output_str.lines().filter(|l| l.contains("=")).count() + bc_blocks,
        exec_steps: rt_steps,
        exec_calls: rt_calls,
    }
}

#[test]
fn snapshots() -> Result<()> {
    let root = Path::new("tests/programs").canonicalize()?;
    for entry in WalkDir::new(&root) {
        let entry = entry?;
        if !entry.file_type().is_file() {
            continue;
        }

        let path = entry.path();
        let ext = path.extension().unwrap().to_str().unwrap();
        if ext != "rice" {
            continue;
        }

        let contents = fs::read_to_string(path)?;
        let first_line = contents.lines().next().unwrap();
        let args = match first_line.strip_prefix("//") {
            Some(args) => args,
            None => "",
        };

        // Compile without optimizations first...
        let output = compile(path, &("".to_string() + args))?;
        let name = path.file_name().unwrap().to_str().unwrap().to_string() + "";
        let snapshot_path = path.parent().unwrap();
        insta::with_settings!({
          snapshot_path => snapshot_path,
          prepend_module_to_snapshot => false,
          filters => vec![
            (r"thread 'main' \((.*?)\)", r"thread 'main' ([some PID])")
          ],
        }, {
          insta::assert_toml_snapshot!(name, output);
        });

        // Compile with optimizations too to ensure they don't affect program behavior.
        let output = compile(path, &(" -O1".to_string() + args))?;
        let name = path.file_name().unwrap().to_str().unwrap().to_string() + ".opt";
        let snapshot_path = path.parent().unwrap();
        insta::with_settings!({
          snapshot_path => snapshot_path,
          prepend_module_to_snapshot => false,
          filters => vec![
            (r"thread 'main' \((.*?)\)", r"thread 'main' ([some PID])")
          ],
        }, {
          insta::assert_toml_snapshot!(name, output);
        });
    }
    Ok(())
}
