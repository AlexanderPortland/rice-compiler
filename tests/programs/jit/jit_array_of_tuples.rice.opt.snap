---
source: tests/snapshots.rs
expression: output
---
success = true
stdout = '''
3
4
1
2
3
'''
stderr = ''
